### ISSUE TITLE
Can't get Hetzner LBs to run with ingress-nginx LB

### ISSUE THREAD

**Issue by stevek-pro** [hetznercloud/hcloud-cloud-controller-manager#257]
Configuring `ingress-nginx controller` with Hetzner LB has been quite the trouble until now. I will describe my setup below the error I am currently facing.

### Error Message at sub.mainhost.de:

```
400 Bad Request
The plain HTTP request was sent to HTTPS port
nginx
```

The **Hetzner Wildcard Certificate** is correctly terminated, I see the lock and inspecting the cert shows `*.mainhost.de`.

I experimented with all ports and combinations tcp, https, http, redirect, wildcard certs, labeling of certs (docs say naming, controller output says labling) & it does not seem to run. 

### Configs

```
cluster $ k describe service -n=ingress-nginx ingress-nginx-controller
Name:                     ingress-nginx-controller
Namespace:                ingress-nginx
Labels:                   app.kubernetes.io/component=controller
                          app.kubernetes.io/instance=ingress-nginx
                          app.kubernetes.io/managed-by=Helm
                          app.kubernetes.io/name=ingress-nginx
                          app.kubernetes.io/version=1.0.4
                          helm.sh/chart=ingress-nginx-4.0.6
Annotations:              load-balancer.hetzner.cloud/certificate-type: managed
                          load-balancer.hetzner.cloud/hostname: mainhost.de
                          load-balancer.hetzner.cloud/http-managed-certificate-domains: mainhost.de
                          load-balancer.hetzner.cloud/http-managed-certificate-name: hcloud-ccm/service-uid:7e46e35f-a1ca-421f-aea6-808cdf0ca482
                          load-balancer.hetzner.cloud/http-redirect-http: false
                          load-balancer.hetzner.cloud/http-sticky-sessions: true
                          load-balancer.hetzner.cloud/ipv6-disabled: true
                          load-balancer.hetzner.cloud/location: nbg1
                          load-balancer.hetzner.cloud/protocol: https
                          load-balancer.hetzner.cloud/use-private-ip: true
Selector:                 app.kubernetes.io/component=controller,app.kubernetes.io/instance=ingress-nginx,app.kubernetes.io/name=ingress-nginx
Type:                     LoadBalancer
IP Family Policy:         SingleStack
IP Families:              IPv4
IP:                       10.104.107.209
IPs:                      10.104.107.209
LoadBalancer Ingress:     mainhost.de
Port:                     http  80/TCP
TargetPort:               http/TCP
NodePort:                 http  30678/TCP
Endpoints:                10.244.6.10:80
Port:                     https  443/TCP
TargetPort:               https/TCP
NodePort:                 https  31625/TCP
Endpoints:                10.244.6.10:443
Session Affinity:         None
External Traffic Policy:  Local
HealthCheck NodePort:     30820
Events:
  Type     Reason                  Age                 From                Message
  ----     ------                  ----                ----                -------
  Warning  SyncLoadBalancerFailed  56m (x3 over 61m)   service-controller  Error syncing load balancer: failed to ensure load balancer: hcloud/loadBalanc
ers.EnsureLoadBalancer: hcops/LoadBalancerOps.ReconcileHCLBServices: port 80 is busy, redirect_http is enabled on another service (source_port_already_us
ed)
  Normal   EnsuredLoadBalancer     54m                 service-controller  Ensured load balancer
  Warning  SyncLoadBalancerFailed  54m (x3 over 54m)   service-controller  Error syncing load balancer: failed to ensure load balancer: hcloud/loadBalanc
ers.EnsureLoadBalancer: hcops/LoadBalancerOps.ReconcileHCLBServices: hcops/LoadBalancerOps.reconcileManagedCertificate: hcops/CertificateOps.CreateManage
dCertificate: cannot create more than 20 managed certificates per hour (resource_limit_exceeded)
  Normal   EnsuringLoadBalancer    54m (x14 over 65m)  service-controller  Ensuring load balancer

```

The Hetzner LB setup for above config is **HTTPS 443 -> 31625** (port on **ingress-nginx controller**) with Hetzner's Wildcard Lets Encrypt Cert for the domain and with http redirect on and off.

I have a webservice on port 80 (and an app with port 3000 behind) with `ingress object`. I tried both with the `ingress object` with cert & without and now playing around with ports. I am using the modern `ingressClassName: nginx`

This is my Ingress Controller:
```

annotations:
    load-balancer.hetzner.cloud/hostname: "mainhost.de" # LB DNS
    load-balancer.hetzner.cloud/location: nbg1
    load-balancer.hetzner.cloud/use-private-ip: "true"
    load-balancer.hetzner.cloud/protocol: "https"
    load-balancer.hetzner.cloud/certificate-type: "managed"
    # load-balancer.hetzner.cloud/http-managed-certificate-domains: "*.mainhost.de"
    # load-balancer.hetzner.cloud/http-managed-certificate-name: "cert:mainhost-wildcard"
    # load-balancer.hetzner.cloud/http-redirect-http: "false"
    load-balancer.hetzner.cloud/http-sticky-sessions: "true"
    load-balancer.hetzner.cloud/ipv6-disabled: "true"
```

It would be super helpful if somebody could draw a little schema how to get this to work with any internal LB but with wildcard certs. I want to have at least the hetzner LB managed an entire domain.

Got it to work directly to a single webservice in my cluster (`type: Loadbalancer`). But I want to use the `ingress-nginx` LB for provisioning its `ingress objects` and LB-ing traffic from Hetzner LB to my microservices (`type: ClusterIP`). 

One should not have to use 20 LB if one has 20 subdomains. This makes no sense from an efficiency standpoint at all. Any help very welcome. The community needs your guidance, Hetzners! 🔢

**Comment by talex-de** [comment]
```
Error syncing load balancer: failed to ensure load balancer: hcloud/loadBalanc
ers.EnsureLoadBalancer: hcops/LoadBalancerOps.ReconcileHCLBServices: hcops/LoadBalancerOps.reconcileManagedCertificate: hcops/CertificateOps.CreateManage
dCertificate: cannot create more than 20 managed certificates per hour (resource_limit_exceeded)
```

You have to take care yourself for the managed certs being created by HCloud LB. I experienced that I have to drop these certs manually by using the Cloud CLI tool. I handle this using Ansible to reinstall or redeploy my Traefik ingress ctrl.

Also, this error tells you that the resource limit has been reached.

Did you check in Hetzner Cloud UI _Project > Security > Certificates_ and what is listed there?

**Comment by talex-de** [comment]
hmm, you try to use an existing cert? @stevek-pro

**Comment by stevek-pro** [comment]
Hi @talex-de Thanks for the tip with the CLI. I didn't try that yet. 

Yes, I'm trying to use existing cert. 

The Kube notifications got preceeded by the latest one, no?

`Normal   EnsuringLoadBalancer    54m (x14 over 65m)  service-controller Ensuring load balancer`

So you are saying your setup works for you with one LB for entire domain and subdomains? Care to share your (Traefik) configuration design?

**Comment by stevek-pro** [comment]
I am also getting in service events:

```
  Warning  SyncLoadBalancerFailed  3m21s (x4 over 8m44s)  service-controller  Error syncing load balancer: failed to ensure load balancer: hcloud/loadBalancers.EnsureLoadBalanc
er: hcops/LoadBalancerOps.ReconcileHCLBServices: hcops/hclbServiceOptsBuilder.buildAddServiceOpts: hcops/CertificateOps.GetCertificateByLabel: not found
```

I declare in service as follows: 

`load-balancer.hetzner.cloud/http-managed-certificate-name: "hcloud-ccm/service-uid: 4ad55eab-ca24-4c2c-8aac-257b05f01921"`

LB is correclty configures, but health check is bad. Are there docs on this?

**Comment by zzorica** [comment]
I think the issue is in the protocol between LB -> k8s-ingress-controller. If using TLS termination on hetzner LB for the backend it should be http for frontend 443 service, as port 80 is auto-redirecting to 443 so that service doesnt talk to k8s-ingress anyway.

Similar issue here I believe https://github.com/hetznercloud/hcloud-cloud-controller-manager/issues/198

**Comment by stevek-pro** [comment]
Punctuation would help me understand what you wrote @zzorica. Anyway, do we ahve a fully working example somewhere, what annotations we need, what config on the ingress controller and what config on the services and ingresses we need?

**Comment by zzorica** [comment]
@stevek-pro sry on that...to me sentences seems just fine :smile: Anyhow I have it working but withous SSL offloading on Hetzner LB

Setup running on:
* k8s 1.21.6
* ingress-nginx (helm chart 4.0.8)
* cert-manager (helm chart 1.6.1)

ingress-nginx
```
      config:
        use-proxy-protocol: "true"

      service:
        enabled: true
    
        annotations:
          load-balancer.hetzner.cloud/name: "k8s-app-lb"
          load-balancer.hetzner.cloud/location: "nbg1"
          load-balancer.hetzner.cloud/uses-proxyprotocol: "true"
          load-balancer.hetzner.cloud/use-private-ip: "true"
          load-balancer.hetzner.cloud/disable-private-ingress: "true"
       
        externalTrafficPolicy: "Local"
        type: LoadBalancer
        external:
          enabled: true
```

grafana (used for test, all other values in a chart are at default)
```
ingress:
      enabled: true
      ingressClassName: nginx
      annotations:
        cert-manager.io/cluster-issuer: letsencrypt-prod
        kubernetes.io/tls-acme: "true"
    
      hosts:
        - grafana.example.com
      extraPaths:
        - path: /*
          pathType: Prefix
          backend:
            service:
              name: ssl-redirect
              port:
                name: use-annotation
      tls:
        - secretName: grafana.example.com
          hosts:
            - grafana.example.com
```

I added couple of more ingresses and all good. But as mentioned here I would like to get SSL offloading/termination setup on Hetzner LB and from there reach k8s ingresses on http port 80.

@stevek-pro hope this helps you some.

**Comment by druppy** [comment]
@zzorica this looks really nice, but how did you setup lets encrypt ?

It seems to me that we need the `dns01` resolve, but I have no idea as to how to make that work in Hetzner DNS.

My issuer setup looks like this

```yaml
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: letsencrypt-prod
  labels:
    name: letsencrypt-prod
spec:
  acme:
    email: drift@moch.dk
    privateKeySecretRef:
      name: letsencrypt-prod
    server: https://acme-v02.api.letsencrypt.org/directory
    solvers:
    - http01:
        ingress:
          class: nginx
```

I get a nice key in my `letsencrypt-prod`, but no certs as there is no way it can verify any challenge.

It seems like you solved this, would you like to share how your issuer are setup.

**Comment by zzorica** [comment]
@druppy must admit I cheated here :blush: I tried dns01 resolver with hetzner but couldn't make it work on first try. So to save some time I just switched to using digitalocean dns for this one
```yaml
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: letsencrypt-prod
spec:
  acme:
    server: https://acme-v02.api.letsencrypt.org/directory
    email: ######
    privateKeySecretRef:
      name: letsencrypt-prod
    solvers:
      - dns01:
          digitalocean:
            tokenSecretRef:
              name: digitalocean-dns
              key: digitalocean_api_token
```

**Comment by druppy** [comment]
Ok, thanks @zzorica, I also found [this](https://github.com/vadimkim/cert-manager-webhook-hetzner) but I was not able to make it work :-(

**Comment by stevek-pro** [comment]
There has been a development in the meanwhile, I discovered https://github.com/vadimkim/cert-manager-webhook-hetzner and am currently working on getting it to run without any luck so far. There is the mention of running a mandatory DNS01 test suite first, which I am now looking into. To make things worse that repo is creating the certificate manually which I did not do since I discovered Let's Encrypt in 2015, so I am really not understanding what is going on there, but I am not ready to give up just yet.
